### PR TITLE
Base level interface for projector and message reducer

### PR DESCRIPTION
**Pull request by blackhatbrigade**

### CODE REVIEW

**Comment by nichoriverae** [comment]
Looks like a great base for moving forward!

File: projector/reducer_config.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package projector
+
+//reducerConfig Contains all of the information needed to use a given reducer.
+type reducerConfig struct {
+	Reducer  MessageReducer
+	PollTime int64
```
Comment (fishybell) [inline projector/reducer_config.go]: This should be of type "time.Duration" ... which is an int64, but makes it clear on how to initialize it (ie. time.Second * 5 or time.Millisecond * 100).

File: projector/messagereducer.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package projector
+
+import (
+	"github.com/blackhatbrigade/gomessagestore"
+)
+
+//MessageReducer Defines the expected behaviours of a reducer that ultimately is used by the projectors.
+type MessageReducer interface {
+	Reduce(msg gomessagestore.Message, previousState interface{})
```
[the remark this answers is not in the document]
Comment (blackhatbrigade) [reply projector/messagereducer.go]: Ooops, haven't gotten there yet with it.  I think this was one of those I fleshed out just enough to get something else done.  I'll fix

File: projector/reducer_config.go (hunk 1 above)
Comment (blackhatbrigade) [reply projector/reducer_config.go]: +1

**Comment by blackhatbrigade** [comment]
I still have a dependency cycle, so hold off of pulling this in.  I'm going to pull Message out to Repository and see if that solves problems.

**Comment by blackhatbrigade** [comment]
Cyclic dependencies fixed.  Not seeing adverse effects from the fix.

File: projector/projector.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+//projector The base supported projector struct.
+type projector struct {
+	repo     repository.Repository
+	reducers []reducerConfig
+}
+
+//CreateProjector Creates a default projector struct that conforms to the interface.
+func CreateProjector(repos repository.Repository) (proj Projector) {
```
Comment (nichoriverae) [inline projector/projector.go]: goes away

**Pull request merged**